jenkins-dsl/jenkins_jira_integration.py

```diff
@@ -0,0 +1,355 @@
+    optional.add_argument('--auto', action='store_true', default='False', help='Update Jira tickets with CI information automatically')
+
+    args = parser.parse_args()
+    global VERBOSE
+    if args.verbose:
+        VERBOSE = True
+
+    # first we build the path to and confirm existence of the buildnum requested
+    jenkins_url = 'https://ci-cassandra.apache.org'
+    log('Connecting to jenkins server...')
+    server = Jenkins(jenkins_url, username=args.jenuser, password=args.jenpass)
+
+    # We store a per-branch JSON file with cached data of test number, previous number, and all failures so we don't
+    # have to query JIRA for each build's data every time we run the script
+    ci_cache = build_local_cache(server, args.branch, args.buildnum)
+
+    log('Retrieving build. Branch: ' + args.branch + '. Build number: ' + args.buildnum)
+    build_data = retrieve_build_details(server, args.branch, args.buildnum)
+
+    # jira_results represents the data we're going to post to the final JIRA ticket about this CI run and test histories
+    jira_results = '[CI Results]\n'
+    jira_results += build_data.string_detailed()
+
+    # Add a space between CI meta and test details for aesthetics
+    jira_results += '\n'
+
+    jira = JIRA('https://issues.apache.org/jira', basic_auth=(args.jirauser, args.jirapass))
+
+    # Putting this in a table goes a long way towards making it parseable
+    jira_results += '||Test|Failures|JIRA||\n'
+    for test_name in build_data.test_failures:
+        failures, total, jiralink = get_test_failure_details(jira, test_name, ci_cache)
+        jira_results += ('|' + test_name + '|' + str(failures) + ' of ' + str(total) + '|' + jiralink + '\n')
+
+    # Next, we want to query JIRA for the cassandra ticket in question and see if we've given it an update yet on build status; protect against spamming
+    post_results_to_jira(jira, build_data.JIRA, args.branch, args.auto, jira_results)
+
+
+class BuildData:
+    def __init__(self, server: jenkins, branch: str, build_num: str) -> None:
+        """
+        A lot of the guts of this are painfully parsed out of the nested data structures in the Python API of a Jenkins build
+        For example, see the Python API link on the following build on trunk: https://ci-cassandra.apache.org/job/Cassandra-trunk/959/testReport/api/
+        (note: you may need to replace 959 w/a more recent build as they fall off the history to see it)
+        :param server: jenkins server to query; expected to already be authenticated and connected
+        :param branch: str of the branch name to pull from
+        :param build_num: str of the build number to query
+        """
+        self.branch = branch
+        self.number = build_num
+
+        self.paths = []
+        # TODO: Determine if depth=5 is necessary or in any way meaningful in this API Query. If so, document why.
+        build = server.get_build_info('Cassandra-' + branch, int(build_num), depth=5)
```

Comment:
Does the python api not have selectors like xml xpath or json tree? (you don't need to be pulling out 50MB+ each time…)

Reply:
I didn't see it: https://python-jenkins.readthedocs.io/en/latest/api.html#jenkins.Jenkins.get_build_info

There's the ability to pass a tree property w/specific things you want from Jenkins but I didn't see that support in the API reference for this lib.

The downside is a big fetch on initial history build + 50MB fetch every N hours as Jenkins jobs complete. Might be worth optimizing to query the API directly instead of leaning on this python wrapper for this query here, but I didn't want to dig into that in V1.

Probably worth a follow up ticket.